File: 2 - Sorting and Order Statistics/2.3 - Sorting in Linear Time/Code/counting_sort.py
"""
counting_sort.py

"""


def counting_sort(array, n, k):
    """Counting sort algorithm

    Time complexity: O(n + k)
    Space complexity: O(n + k)

    Args:
        array (list[int]): Input array
        n (int): Number of elements in the input array
        k (int): Maximum value of the input array

    Returns:
        list[int]: Sorted array
    """
    sorted_output = [0] * n
    memo = [0] * (k + 1)

    for j in range(n):
        memo[array[j]] = memo[array[j]] + 1

    for i in range(1, k + 1):
        memo[i] = memo[i] + memo[i - 1]

    for x in range(n - 1, -1, -1):
        sorted_output[memo[array[x]] - 1] = array[x]
        memo[array[x]] = memo[array[x]] - 1

    return sorted_output

File: 2 - Sorting and Order Statistics/2.3 - Sorting in Linear Time/Code/CountingSort.js
/**
 * CountingSort.js
 *
 */

function countingSort(array, n, k) {
    let sortedOutput = Array(n).fill(0);

    let memoLength = k + 1;
    let memo = Array(memoLength).fill(0);

    for (let i = 0; i < n; i++) {
        memo[array[i]] = memo[array[i]] + 1;
    }

    for (let j = 1; j <= k; j++) {
        memo[j] = memo[j] + memo[j - 1];
    }

    for (let x = n - 1; x > -1; x--) {
        sortedOutput[memo[array[x]] - 1] = array[x];
        memo[array[x]] = memo[array[x]] - 1;
    }

    return sortedOutput;
}

module.exports = countingSort;
